sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -246,6 +247,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(projectList, Join(pruneJoinChild(left), pruneJoinChild(right), joinType, condition))
 
+    // Eliminate unneeded attributes from either side of a Join.
+    case Project(projectList, Filter(predicates, Join(left, right, joinType, condition)))  =>
```

Comment:
I think this problem shouldn't occur in Aggregate since it's covered by the pattern match:
case a @ Aggregate(_,_,child) if (child.outputSet -- a.references).nonempty.
As for Sort, I still need to think it over.